### ISSUE TITLE
docx output is corrupted if "date" field only includes numbers

### ISSUE THREAD

**Issue by jkr** [jgm/pandoc#3015]
As reported on the mailing list:

The following md produces corrupted docx output

```{.markdown}
---
date: 20160706
---

foo
```

Note that this is also corrupted with `date: 16-07-06` but **not** with `date: 2016-07-06`. Other separators (slashes, periods, spaces) do not produce corruption. Adding in letters also doesn't produce corruption.

Also: this doesn't produce corrupt output in `title` or `author` field -- just `date`.

Weird, right?

**Comment by jkr** [comment]
Okay -- the problem seems to be corrupted `<dcterms:created>` date at line 446 ion `docProps/core.xml`. So it's something about the `normalizeDate $ stringify $ docDate meta` step.

**Comment by jkr** [comment]
(I mean, of course, line 446 in the writer, writing to `docProps/core.xml`)

**Comment by jkr** [comment]
@jgm: the issue here is that when `normalizeDate` thinks it can do the right thing, but can't. e.g. the various corrupted numerical formulas above. Even if it does give a non-corrupting value, it might well be wrong (if you use, say, `%y-%m-%d`). I guess my suggestion would be to skip this step and either take the real time of creation (`getCurrentTime`, since `writeDocx` is in IO, and we already use it for fallback on track-changes) or just skip it altogether, as if `normalizeDate` had returned a `Nothing`. Thoughts?

**Comment by jgm** [comment]
+++ Jesse Rosenthal [Jul 06 16 07:54 ]:
>   [1]@jgm: the issue here is that when normalizeDate thinks it can do the
>   right thing, but can't. e.g. the various corrupted numerical formulas
>   above. Even if it does give a non-corrupting value, it might well be
>   wrong (if you use, say, %y-%m-%d). I guess my suggestion would be to
>   skip this step and either take the real time of creation
>   (getCurrentTime, since writeDocx is in IO, and we already use it for
>   fallback on track-changes) or just skip it altogether, as if
>   normalizeDate had returned a Nothing. Thoughts?

The other writers all use the date given in the document,
so I think the docx writer should too.  So what does
normalizeDate do with this input?  Does it just take the
whole number as the year, and Word rejects that?  I suppose
we coud do some kind of sanity check in normalizeDate
and return Nothing if it fails...

**Comment by jkr** [comment]
A sanity-check would probably do the trick. It looks like the problem is that `parseTimeM` will accept anything as a `%Y`. So we get weird results like: 

~~~
"76-03-18"  ->  "0076-03-18"
"19760318"  ->  "19760318-01-01"
~~~

These are out of word's allowable year range, which trial-and-error shows to be 1601-9999 inclusive. (The [ISO 8601 standard](https://en.wikipedia.org/wiki/ISO_8601#Years) says you're supposed to start at 1583, so MS isn't quite being standards-compliant here.)

So I guess I'd suggest adding a couple more formats (`"%Y%m%d", "%Y%m"`) before `"%Y"` in the formats list to catch some of the things that give us long numbers. Then we could do a sanity-check on the result and either `Nothing` out if the year is out of the range or try to do something smart and guess (if less than 100 add 2000, etc). Probably better to return `Nothing` than guess wrong, though, so I'd go with adding formats and then just rejecting every year out of range..

**Comment by jkr** [comment]
(Just to confirm: none of this has to do with the date as it's written in the document -- this is just about the metdata entered in `docProps/core.xml`)

**Comment by jkr** [comment]
@jgm how about his: https://github.com/jkr/pandoc/tree/legalDates

It does as described above. Add a couple of formats to catch longer strings of numbers. And then disallow years out of the 1601-9999 range. It seems to solve the problem that opened this issue. I'll push if it looks good to you.

**Comment by jgm** [comment]
I think I'm fine with it, but I had some doubts after your
last message

>    (Just to confirm: none of this has to do with the date
>    as it's written in the document -- this is just about the metdata entered
>    in docProps/core.xm

Maybe it *would* be better just to use the date the docx
is being created (via pandoc)?  I suppose one question is
whether your docx reader parses this date (in
docProps/core.xml) and uses it to populate the metadata
date field.  If so, that would be a good reason to try
to populate the docProps/core.xml date from the metadata
date in the writer.  If not, maybe we should just go for
the simpler thing?  I'm not really sure what hangs on this.

+++ Jesse Rosenthal [Jul 09 16 12:44 ]:

**Comment by jkr** [comment]
John MacFarlane <notifications@github.com> writes:

> Maybe it *would* be better just to use the date the docx
> is being created (via pandoc)?  I suppose one question is
> whether your docx reader parses this date (in
> docProps/core.xml) and uses it to populate the metadata
> date field.

Nope -- I don't use it. There can be a lot of circumstances where
there's a mismatch. If I'm writing a conference paper or a lecture, I'll
post-date the document so it will reflect the day I deliver it, but the
creation date or modification date will often be earlier.

I don't think we get much out of putting the user-supplied date in
`docProps/core.xml`. What we seem to end up in the properties window of
the document is the claim that it was created originally at the
user-supplied date, and modified at the real time of creation (I'm not
sure where that date is kept). This could produce some weird results
with post-dating or mis-parsing.

(Also, because of that modification date, we can't guarantee that you'd
have the ability to lie about metadata for the sake of backing up a
"forgot-to-attach-it" alibi. Students would have to do what they already
do, and change their computer time if they want their metadata to lie.)

I still think the changes I suggested are good, because if we're going
to get a "YYYY-MM-DD", we might as well get a legal one, and starting at
1601 seems like a safe way to keep it ISO-8601-compliant (who knows what
other busymody xml formats might decide to validate?).

So maybe push the legalDate changes, and in addition switch the writer
to using `getCurrentTime` for its creation date? We already pull it in
as a local value for writing in track-changes info, so we don't have to
do anything. I'll just do a bit of digging to see whether it's zoned or
UTC.

**Comment by jkr** [comment]
Okay -- I pushed the latter change to the writer in the [legalDates branch](https://github.com/jkr/pandoc/commits/legalDates). I suppose if users *really* wanted to be able to lie about document props, the writer could read some sort of specialized YAML metadata

~~~{.yaml}
docx-props:
    created: 2016-07-06T14:00:00
    modified: etc...
~~~

But again, it would be difficult to guarantee that this would do what people wanted.

**Comment by jkr** [comment]
@jgm : okay with you if I push these changes? Pretty innocuous, but since `normalizeDate` touches a few other writers, I wanted to check.

**Comment by jgm** [comment]
Yes, go ahead.

**Comment by jkr** [comment]
closed with f529fc17efff60b5ba32cb8c82aef356a84c3d3d
